command/agent/command.go

```diff
@@ -427,10 +432,130 @@ func (c *Command) startAgent(config *Config, agent *Agent,
 	if config.Discover != "" {
 		c.Ui.Info(fmt.Sprintf("  mDNS cluster: %s", config.Discover))
 	}
+
 	return ipc
 }
 
 // startupJoin is invoked to handle any joins specified to take place at start time
+func (c *Command) startupJoinSRV(config *Config, agent *Agent) error {
+	if len(config.JoinSRV) == 0 {
+		return nil
+	}
+
+	c.Ui.Output(fmt.Sprintf("Joining cluster via SRV...(replay: %v)", config.ReplayOnJoin))
+	n, err := c.joinSRV(agent, config.ReplayOnJoin, config.JoinSRV)
+	if err != nil {
+		return err
+	}
+
+	if n == 0 {
+		return errors.New("Failed to join any hosts via SRV")
+	}
+
+	c.Ui.Info(fmt.Sprintf("Join completed. Synced with %d initial agents", n))
+	return nil
+
+}
+
+// retryJoinSRV is invoked to handle joins with retries. This runs until at least a
+// single successful join or RetryMaxAttempts is reached
+func (c *Command) retryJoinSRV(config *Config, agent *Agent, errCh chan struct{}) {
+	// Quit fast if there is no nodes to join
+	if len(config.RetryJoinSRV) == 0 {
+		return
+	}
+
+	// Track the number of join attempts
+	attempt := 0
+	for {
+		// Try to perform the join
+		n, err := c.joinSRV(agent, config.ReplayOnJoin, config.RetryJoinSRV)
+
+		if err != nil {
+			c.logger.Printf("[ERR] agent: Failed to join via SRV: %v", err)
+		}
+
+		if err == nil && n > 0 {
+			c.logger.Printf("[INFO] agent: Join completed. Synced with %d initial agents", n)
```

Comment:
I see. Sorry for my confusion I just want to understand how this will be used and how it differs from existing options.

In your example, what kind of process decides who gets added to the SRV record? If you always populated the SRV record with all the Consul servers, for example, it would always be sufficient for new nodes to join.

I hope I haven't yagni-d you with the join options but that's where I thought it was going :-) For the polling case I'd had a -discover_srv option that takes the list and calls your function forever in a separate go routine vs. -1 retry count. If you can use the new join options without adding the polling support I still think that's ideal, though.

> On Sep 15, 2015, at 6:35 AM, Dale Hamel <notifications@github.com> wrote:
> 
> In command/agent/command.go:
> 
> > +	if len(config.RetryJoinSRV) == 0 {
> > +		return
> > +	}
> > +
> > +	// Track the number of join attempts
> > +	attempt := 0
> > +	for {
> > +		// Try to perform the join
> > +		n, err := c.joinSRV(agent, config.ReplayOnJoin, config.RetryJoinSRV)
> > +
> > +		if err != nil {
> > +			c.logger.Printf("[ERR] agent: Failed to join via SRV: %v", err)
> > +		}
> > +
> > +		if err == nil && n > 0 {
> > +			c.logger.Printf("[INFO] agent: Join completed. Synced with %d initial agents", n)
> That assumes that the SRV record points to the hosts in the existing cluster. If that changes, you could have a split where you have two clusters who will never know about each other.
> 
> I think I mentioned this concern in another comment.
> 
> Let's say you have 4 nodes, A,B,C,D. Only A and B are in the SRV record. C and D join them.
> 
> A and B get torn down, and a new SRV record is built with E and F. You now have two clusters, DC, EF. If DC keep checking the SRV record ,they will find EF. If they don't, you'll have a partition.
> 
> —
> Reply to this email directly or view it on GitHub.
>